skbio/io/format/embedding.py

```diff
@@ -0,0 +1,212 @@
+                return True, {}
+
+    return False, {}
+
+
+@embed.reader(None)
+def _embed_to_generator(
+    fh,
+    constructor=ProteinEmbedding,
+    obj_constructor=Protein,
+    kwargs: dict = {},
+):
+    h5grp = h5py.File(fh, "r")
+    embed_fh = h5grp["embedding"]
+    id_fh = h5grp["id"]
+    idptr_fh = h5grp["idptr"]
+    j = 0
+    n = idptr_fh.shape[0]
+    for i in range(n):
+        idptr = idptr_fh[i]
+        id_ = id_fh[j:idptr]
+        emb = embed_fh[j:idptr]
+        string = str(id_.tobytes().decode("ascii")).replace("\x00", "")
+        j = idptr
+        yield constructor(emb, string, **kwargs)
+
+
+def _embed_to_object(
+    fh,
+    constructor=ProteinEmbedding,
+    obj_constructor=Protein,
+    kwargs: dict = {},
+):
+    h5grp = h5py.File(fh, "r")
+    embed_fh = h5grp["embedding"]
+    id_fh = h5grp["id"]
+
+    # assumes that there is only a single object in the file
+    emb = embed_fh[()].squeeze()
+    id_ = id_fh[()]
+    string = str(id_.tobytes().decode("ascii")).replace("\x00", "")
+    return constructor(emb, string, **kwargs)
+
+
+@embed.reader(ProteinEmbedding)
+def _embed_to_protein(
+    fh,
+    kwargs: dict = {},
+):
+    return _embed_to_object(
+        fh,
+        constructor=ProteinEmbedding,
+        obj_constructor=Protein,
+        kwargs=kwargs,
+    )
+
+
+def _objects_to_embed(objs, fh):
+    with h5py.File(fh, "w") as h5grp:
+        for i, obj in enumerate(objs):
+            # store string representation of the object
+            # that will serve as an identifier for the entire object.
+            # for sequences, this could be the sequence itself
+            # for molecules, this could be the SMILES string.
+            # The entries in this string representation can be used
+            # to index the row vectors in the embedding.
+            # For sequences, this is the positional index of the sequence.
+            # For molecules, this is the position index of atoms in the SMILES string.
+            arr = np.frombuffer(str(obj).encode("ascii"), dtype=np.uint8)
+            # Store the embedding itself. We are assuming that the
+            # embbedding is a 2D numpy array
+            emb = obj.embedding
+            # store the pointers that keep track of the start and
+            # end of the embedding for each object, as well as well as
+            # the corresponding string representation
+            if "idptr" in h5grp:
+                idptr_fh = h5grp["idptr"]
+                idptr_fh.resize((i + 1,))
+                idptr_fh[i] = len(arr) + idptr_fh[i - 1]
+            else:
+                idptr_fh = h5grp.create_dataset(
+                    "idptr", data=[len(arr)], maxshape=(None,), dtype=np.int32
+                )
+            if "id" in h5grp:
+                id_fh = h5grp["id"]
+                id_fh.resize((idptr_fh[i],))
+                id_fh[idptr_fh[i - 1] : idptr_fh[i]] = arr
+            else:
+                id_fh = h5grp.create_dataset(
```

Comment:
Should any of these have compression?

Reply:
The sequences, probably not worth it.  The embedding arrays possibly yes -- since that is where most of the memory footprint is going to be coming from.